src/main/java/com/faforever/client/player/UsernameChangeController.java

```diff
@@ -0,0 +1,73 @@
+package com.faforever.client.player;
+
+import com.faforever.client.fx.Controller;
+import com.faforever.client.notification.NotificationService;
+import com.faforever.client.user.UserService;
+import javafx.beans.property.ReadOnlyStringProperty;
+import javafx.beans.value.ChangeListener;
+import javafx.event.ActionEvent;
+import javafx.scene.Node;
+import javafx.scene.control.Button;
+import javafx.scene.control.TextField;
+import javafx.scene.layout.VBox;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+import javax.inject.Inject;
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class UsernameChangeController implements Controller<Node> {
+  private final UserService userService;
+  private final NotificationService notificationService;
+  public TextField usernameField;
+  public VBox root;
+  public Button changeButton;
+  private Runnable callback;
+  private ReadOnlyStringProperty displayPlayer;
+
+  @Inject
+  public UsernameChangeController(UserService userService, NotificationService notificationService) {
+    this.userService = userService;
+    this.notificationService = notificationService;
+  }
+
+  @Override
+  public void initialize() {
+    usernameField.textProperty().addListener((observable, oldValue, newValue) -> changeButton.setDisable(displayPlayer == null || newValue.equals(displayPlayer.get())));
+  }
+
+  public void setDisplayPlayer(ReadOnlyStringProperty displayPlayer) {
+    this.displayPlayer = displayPlayer;
+    ChangeListener<String> changeListener = (observable, oldValue, newValue) -> {
+      if (!userService.getUsername().equals(newValue)) {
+        root.setVisible(false);
+        return;
+      }
+      usernameField.setText(newValue);
+    };
+    changeListener.changed(displayPlayer, null, displayPlayer.get());
+    displayPlayer.addListener(changeListener);
```

Comment:
Whenever you work with listeners, think of memory leaks. As a rule of thumb, whenever a short-lived object adds a listener to a long-lived object, you create a memory leak.

`UsernameChangeController` is of scope prototype, therefore a new bean is created every time needed. It's therefore a short-lived object.
The player object, to which the listener is added, is a longed-lived object. If you create 10 user change controllers, the player object will keep a reference to all of them, preventing them from being garbage collected.

One solution to this problem is to remove the listener when the controller is being destroyed, but that's rather inconvenient.
Another solution, already used in the client, is to use weak references. Weak references do not prevent objects from being garbage collected. But then you need to prevent the listener from being collected too early.

However, I'm not sure if we even need a change listener here. What's the exact use case?